src/main/java/org/broadinstitute/hellbender/tools/HtsgetReader.java

```diff
@@ -130,161 +107,74 @@
         }
     }
 
-    private ObjectMapper getObjectMapper() {
-        final ObjectMapper mapper = new ObjectMapper();
-        mapper.enable(DeserializationFeature.UNWRAP_ROOT_VALUE);
-        mapper.configure(MapperFeature.ACCEPT_CASE_INSENSITIVE_PROPERTIES, true);
-        return mapper;
-    }
-
     @Override
     public Object doWork() {
-        // construct request from command line args and convert to URI
-        final HtsgetRequestBuilder req = new HtsgetRequestBuilder(endpoint, id)
+        // Construct request from command line args and convert to URI
+        final HtsgetRequest req = new HtsgetRequest(endpoint, id)
             .withFormat(format)
             .withDataClass(dataClass)
             .withInterval(interval)
             .withFields(fields)
             .withTags(tags)
             .withNotags(notags);
-        final URI reqURI = req.toURI();
-
-        final HttpGet getReq = new HttpGet(reqURI);
-        try (final CloseableHttpResponse resp = this.client.execute(getReq)) {
-            // get content of response
-            final HttpEntity entity = resp.getEntity();
-            final Header encodingHeader = entity.getContentEncoding();
-            final Charset encoding = encodingHeader == null 
-                ? StandardCharsets.UTF_8
-                : Charsets.toCharset(encodingHeader.getValue());
-            final String jsonBody = EntityUtils.toString(entity, encoding);
-
-            final ObjectMapper mapper = this.getObjectMapper();
 
-            if (resp.getStatusLine() == null) {
-                throw new UserException("htsget server response did not contain status line");
-            }
-            final int statusCode = resp.getStatusLine().getStatusCode();
-            if (400 <= statusCode && statusCode < 500) {
-                final HtsgetErrorResponse err = mapper.readValue(jsonBody, HtsgetErrorResponse.class);
-                throw new UserException("Invalid request, received error code: " + statusCode + ", error type: "
-                        + err.getError() + ", message: " + err.getMessage());
-            } else if (statusCode == 200) {
-                final HtsgetResponse response = mapper.readValue(jsonBody, HtsgetResponse.class);
-
-                if (this.readerThreads > 1) {
-                    this.getDataParallel(response);
-                } else {
-                    this.getData(response);
-                }
-
-                logger.info("Successfully wrote to: " + outputFile);
+        final HtsgetResponse resp = req.getResponse();
+        if (resp.getMd5() == null) {
+            this.checkMd5 = false;
+            logger.info("No md5 checksum received");
+        }
 
-                if (checkMd5) {
-                    this.checkMd5(response);
-                }
+        try (final OutputStream outputstream = new FileOutputStream(this.outputFile)) {
+            if (this.parallelDownload) {
+                resp.getBlocks().parallelStream().map(block -> {
```

Comment:
this is very clean but we might want to keep the ability to tune by number of threads

Reply:
have you seen that java actually runs the map operations in parallel and then rejoins them in the right oder?  I'm not 100% sure that it's guaranteed to do either.

Reply:
I'll revert the change with the threads. Whether the stream is ordered/not ordered is separate from whether it's parallel/sequential, so the operations might run in parallel and out of order, but the `forEachOrdered` is makes sure they wait in order for the final step.